### ISSUE TITLE
[clang static analyzer] core.NullDereference false negative with `*p = 1024`

### ISSUE THREAD

**Issue by hallo-wereld** [llvm/llvm-project#62232]
version: Clang trunk
args: 
```shell
--analyze -Xclang -analyzer-stats -Xclang -analyzer-checker=core,debug.ExprInspection --analyzer-output text
```

This case reports `core.NullDereference`.
```c
int foo(int t)
{
    if (!t)
    {
        int *p = (void *)t;
        *p = 1024;
    }
}
```
This case doesn't report `core.NullDereference`.
```c
int foo(int t)
{
    t = t + 1;
    if (!t)
    {
        int *p = (void *)t;
        *p = 1024;
    }
}
```

See it live: 
https://godbolt.org/z/j8nMonozq
https://godbolt.org/z/Yzhdb7KEx

**Comment by steakhal** [comment]
We create a `SymExpr` representing the calculation, like `z + 3`, then we bind it to `x`. After this, we introduce a constraint that `z + 3 is zero`. Subsequent loads of variable `x` should fold/simplify into a `ConcreteInt(0)`, which can be cast later to anything, and would do the right thing.

Now the question is, why doesn't the load of `x` not simplify?
My guess would be that `RangeConstraintManager::getSymVal()` somehow doesn't realize that this has a single possible value.
This doesn't look like a good first issue, but it's the easier side of the constraint manager issues for sure.

**Comment by danix800** [comment]
For `x = z + 3`, `x` is bound to symbol value `z + 3`. Before assumption on the `if` branch, `!x` is transformed into `z == -3`, which gives a constraint that `z` &in; `[-3, -3]`, the assumption result is true, but no further processing for `x` for now.

`x` can be simplified further before usual arithmetic on `NonLoc`, but not for bitcast like `(void *)x` (from `NonLoc` to `Loc`), this results in an `UnknownVal`.

For `x = z * 3`,  `x` is bound to symbol value `z * 3`, no transformation before assumption. The binary symbol value
`z * 3` is `0`, which is a `ConcreteInt`. The subsequent casts results in a nullptr.
